src/Entity/Citation.php

```diff
@@ -0,0 +1,377 @@
+      ->setSetting('is_ascii', TRUE)
+      ->setSetting('max_length', FieldStorageConfig::NAME_MAX_LENGTH);
+
+    return $fields;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getParentEntity() {
+    if (!isset($this->get('parent_type')->value) || !isset($this->get('parent_id')->value)) {
+      return NULL;
+    }
+
+    $parent = $this->entityTypeManager()
+      ->getStorage($this->get('parent_type')->value)
+      ->load($this->get('parent_id')->value);
+
+    // Return current translation of parent entity, if it exists.
+    if (
+      $parent != NULL &&
+      ($parent instanceof TranslatableInterface) &&
+      $parent->hasTranslation($this->language()->getId())
+    ) {
+      return $parent->getTranslation($this->language()->getId());
+    }
+
+    return $parent;
+  }
+
+  /**
+   * {@inheritdoc}
+   */
+  public function setParentEntity(ContentEntityInterface $parent, $parent_field_name) {
+    $this->set('parent_type', $parent->getEntityTypeId());
+    $this->set('parent_id', $parent->id());
+    $this->set('parent_field_name', $parent_field_name);
+    return $this;
+  }
+
+  /**
+   * {@inheritDoc}
+   */
+  public function getBibliography($style = self::APA): string {
+
+    $data = [
+      'id' => $this->id(),
+      'title' => $this->label(),
+      // Custom variables that wrap the title with a link tag.
+      'link-beginning' => $this->getLinkBeginning(),
+      'link-ending' => $this->getLinkBeginning() ? '</a>' : NULL,
+      'DOI' => $this->getDoi(),
+      'URL' => $this->getUrl(),
+      'author' => $this->getAuthor(),
+      'edition' => (int) $this->getEdition(),
+      'issue' => $this->getIssue(),
+      'issued' => $this->getDate(),
+      'genre' => $this->getGenre(),
+      'page' => $this->getPage(),
+      'publisher' => $this->getPublisher(),
+      'publisher-place' => $this->getPublisherPlace(),
+      'subtitle' => $this->getSubtitle(),
+      'type' => $this->getType(),
+      'volume' => (int) $this->getVolume(),
+    ];
+
+    if ($data['type'] == 'article-journal') {
+      $data['collection-title'] = $data['publisher'];
+    }
+
+    // Convert the arrays into objects.
+    $data = json_decode(json_encode([array_filter($data)]));
+
+    $local_csl = __DIR__ . '/Styles/' . $style . '.xml';
+    if (!file_exists($local_csl)) {
+      return '';
+    }
+
+    // Load the style CSL file.
+    $style = file_get_contents($local_csl);
+    $citeProc = new CiteProc($style);
+    return htmlspecialchars_decode($citeProc->render($data));
+  }
+
+  /**
+   * If the citation can be linked to a url, get the first part of the <a> tag.
+   *
+   * @return string|null
+   *   First half of the <a> tag.
+   *
+   * @throws \Drupal\Core\Entity\EntityMalformedException
+   */
+  protected function getLinkBeginning() {
+    if ($link = $this->getLink()) {
+      // This will pull out the `<a href....>` part of the link.
+      preg_match('/<.*?>/', (string) $link->toString(), $matches);
```

Comment:
```suggestion
      preg_match('/<a.*?>/', (string) $link->toString(), $matches);
```

Reply:
Why not lock it to the `a` tag?

Reply:
I already committed your suggestion yesterday. :+1: nothing to see here.